experimental/dds/tree2/src/codec/parameterized-dds.md

```diff
@@ -0,0 +1,252 @@
+
+### Problems
+
+Consider now a case where the DDS authors wants to introduce parameters which impact the format.
+A natural option would be to augment `create` and `getFactory`:
+
+```typescript
+interface SharedSomethingOptions {
+	formatVersion: number;
+}
+
+class SharedSomething extends SharedObject {
+	public static create(
+		runtime: IFluidDatastoreRuntime,
+		id?: string,
+		params?: Partial<SharedSomethingOptions>,
+	): SharedSomething {
+		// Uh oh--params unused!
+		return runtime.createChannel(id, SharedSomethingFactory.Type) as SharedSomething;
+	}
+
+	public static getFactory(params?: Partial<SharedSomethingOptions>): SharedSomethingFactory {
+		return new SharedSomethingFactory(params);
+	}
+}
+```
+
+This immediately uncovers one issue: since the static `create` method isn't in charge of constructing the DDS's factory,
+there's no place to plumb through the parameters.
+Even worse, the container author likely already wrote some code like this:
+
+```typescript
+const myDataObject = new DataObjectFactory("foo", MyDataObject, [
+	SharedSomething.getFactory({ formatVersion: 5 }),
+]);
+```
+
+If they later write `SharedSomething.create(runtime, { formatVersion: 4 })`, the create flow outlined above will instead use the
+factory created at data object registry construction time.
+
+There is another, more subtle issue with this strategy: how should one implement a `SharedSomethingFactory` which takes in some parameters?
+
+```typescript
+class SharedSomethingFactory {
+	public constructor(private readonly options: SharedSomethingOptions) {}
+
+	public load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		channelAttributes: Readonly<IChannelAttributes>,
+	): Promise<IChannel> {
+		const sharedSomething = new SharedSomething(
+			runtime,
+			id,
+			this.options /* uh oh: channelAttributes and this.options might disagree on snapshot format! */,
+		);
+		await sharedSomething.load(services);
+		return sharedSomething;
+	}
+
+	public create(runtime: IFluidDataStoreRuntime, id: string): IChannel {
+		const sharedSomething = new SharedSomething(document, id, this.options);
+		sharedSomething.initializeLocal();
+		return sharedSomething;
+	}
+}
+```
+
+If the channel's attributes clearly reflect its parameters (e.g. the snapshot format being used),
+then the resulting factory sometimes produces DDSes which don't match the requested format!
+
+There are essentially two ways to resolve this issue:
+
+1. Make `SharedSomethingFactory` have a parameterless constructor, and pass parameters all the way through the `create` flow (i.e. through `createChannel`, factory lookup, and factory's `create` method).
+2. Keep parameters in `SharedSomethingFactory`'s constructor, but change their semantics such that they only necessarily apply to newly created documents.
```

Comment:
If we think customers will want to choose different parameters on a per-instance basis, I think it's better for those parameters to be provided on a per-instance instantiation basis.  If the number of options is large, I don't want to have to configure N! factories.

Reply:
One main goal I have here is to provide better evolvability options on the parameters. I think the vast majority of parameters will generally have one value at a given time that the container author would like to use (e.g. "format version" is usually one such option). Attribution is a case where I could easily imagine a container author wants on only some types of objects--perhaps a text document where the title and the body are in different SharedStrings, and the app doesn't want to display any UI for who changed the title. If we implement other pay-to-play features, I still think it'd be unlikely apps experience combinatorial explosion (and if so, they must be somehow using all of those combinations of features in strange ways...)

Back to evolvability--as soon as an app has multiple ways they want to parameterize a DDS within the same data store, I think it'll be very difficult to reason about evolvability within that data store without names for the different sets of parameters. E.g. say I have an app like above, with registry entries `[["title", SharedString.getFactory()], ["body", SharedString.getFactory({ attributionVersion: 1 })]]` and FF wants to implement some protocol/compat-breaking changes to the way attribution data is stored. After the app does dark rollout of the FF code that understands this change, they need to somehow declare that "OK, we're good to use version 2." Of course, they only want to use version 2 on the things that were already previously using version 1--not the things that didn't have attribution on at all! In the scheme where we have a registry like above, they would just be able to update that registry to `[["title", SharedString.getFactory()], ["body", SharedString.getFactory({ attributionVersion: 2 })]]`. In a scheme where factories don't get parameters on instantiation and are meant to realize whatever parameters/attributes they got from create/load, things get a lot more messy.

Reply:
the real problem is data at rest. you can never assume all document that exist have been opened, so you can never reduce your number of named entries, as those names are persisted. This is another reason i prefer something more like dependency injection.